### PR TITLE
MAILET-115 Refactor AbstractRedirect & AbstractNotify based mailets

### PR DESCRIPTION
**Pull request by aduprat**
#423 + commits (reverse order)
30eca266b58fced958bf9025a6ac0809e39f0b61 Antoine Duprat MAILET-115 Convert RedirectNotify to interface
8d3ec7a5175a9840880c2687a230fd5af548d74b Antoine Duprat MAILET-115 Extract arrayToString from GenericMailet
45f3e3dc5ec35d01a779cb170f6747edd9205d54 Antoine Duprat MAILET-115 Rename AbstractRedirect
96e933c2ecfa48bccec0c856f2a3da3b62ffe726 Antoine Duprat MAILET-115 Remove abstract service with original mail implementation
f08c0920935de31b14e4501331bd1f233697ba61 Antoine Duprat MAILET-115 Remove abstract getSender with original mail implementation
9a85eae1e8e727d1d97bdbe6b3f8d1b6d6fc94b0 Antoine Duprat MAILET-115 Remove abstract getReplyTo with original mail implementation
640d26863f27cb19cab9646a20ee376ec7deddb2 Antoine Duprat MAILET-115 Remove abstract getTo with original mail implementation
60a25af598c7af248e15750a7a4b682c87b411f7 Antoine Duprat MAILET-115 Move message altering from AbstractRedirect to AlteredMailUtils
cd99204aee431f5d1ef1038cb771d2bac853a6cd Antoine Duprat MAILET-115 Move set methods from AbstractRedirect to MailModifier
231fb53b7c3e41a2be41773d01949185336136b6 Antoine Duprat MAILET-115 Redirect & Notify mailets should be responsible of subject modification
fd4f07984cc12b8586288e28e6b6e05ac7c27420 Antoine Duprat MAILET-115 Extract subject prefixing from MimeMessageModifier to MimeMessageUtils
1aee99324173f460cb24bf6ddda50b367414a262 Antoine Duprat MAILET-115 Convert buildNewSubject static method to private method
250e6b1742c1b67d0b62ee351c8cced459906e3e Antoine Duprat MAILET-115 Add tests on subject modification/creation
ac29a5f3c1c4592b901c99d94446ef8c526673f6 Antoine Duprat MAILET-115 Remove abstract getRecipients with original mail implementation
c91be67e75e1339f7082ca58407112722d6ac281 Antoine Duprat MAILET-115 Refactor getSender & getReversePath
a522a9f0151d841d36c9148c0185ae31283fdfb3 Antoine Duprat MAILET-115 Remove abstract getSender implementation
446b6b0f726211fe972c3372257040f326c7f5b7 Antoine Duprat MAILET-115 Extract isUnalteredOrReversePathOrSender from AbstractRedirect to MailAddressesUtils
5b6d522242b81c8670d357d432988bd4eac5b459 Antoine Duprat MAILET-115 Remove abstract getReversePath implementation
bfc2e0a2323b7f99229dad117bb74ef84e24b1bd Antoine Duprat MAILET-115 Remove abstract getRecipients implementation
2139a8b64b455e0b38b70abca6b26dc92fecafd1 Antoine Duprat MAILET-115 Use MailAddressUtils and use List instead of arrays/sets
5dabf06cfbcbc7581863d33087b305ea94099244 Antoine Duprat MAILET-115 Introduce mail address utility to transform InternetAddresses/MailAddresses
abb1d2712da7358b67a0b378363247bceaf56b7e Antoine Duprat MAILET-115 Extract replaceInternetAddresses from AbstractRedirect to SpecialAddressesUtils
6f63674ce97d8bf97ac2e49beeeffc4b892666a6 Antoine Duprat MAILET-115 getRecipients should return a list
c16860c9228cbb065473001b185922e03601830c Antoine Duprat MAILET-115 Extract replaceMailAddresses from AbstractRedirect to MailAddressesUtils
9153f0ffc40d7e070fb3fea547659611f11422b0 Antoine Duprat MAILET-115 Moving subject prefixing from AbstractRedirect to MimeMessageModifier

### CODE REVIEW

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java [hunk 1]
```diff
@@ -45,43 +52,154 @@
...
+        mimeMessage.setSubject("My subject");
+        FakeMail mail = FakeMail.builder()
+                .name(MAILET_NAME)
+                .sender(MailAddressFixture.ANY_AT_JAMES)
+                .mimeMessage(mimeMessage)
+                .build();
+
+        bounce.service(mail);
+
+        assertThat(mail.getMessage().getSubject()).isEqualTo("pre My subject");
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: You should check this on the SentMail IMO

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java [hunk 2]
```diff
@@ -45,43 +52,154 @@
...
+    }
 }
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: You should also check what happens when I have a message passed by configuration of Bounce

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        FakeMail mail = FakeMail.builder()
+                .name(MAILET_NAME)
+                .sender(MailAddressFixture.ANY_AT_JAMES)
+                .mimeMessage(mimeMessage)
+                .build();
+
+        forward.service(mail);
+
+        assertThat(mail.getMessage().getSubject()).isEqualTo(expectedSubject);
+    }
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java]: Check that mail is sent

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifyPostmasterTest.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        FakeMail mail = FakeMail.builder()
+                .name(MAILET_NAME)
+                .sender(MailAddressFixture.ANY_AT_JAMES)
+                .mimeMessage(mimeMessage)
+                .build();
+
+        notifyPostmaster.service(mail);
+
+        assertThat(mail.getMessage().getSubject()).isEqualTo("pre My subject");
+    }
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifyPostmasterTest.java]: Check that a mail is sent

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        FakeMail mail = FakeMail.builder()
+                .name(MAILET_NAME)
+                .sender(MailAddressFixture.ANY_AT_JAMES)
+                .mimeMessage(mimeMessage)
+                .build();
+
+        notifySender.service(mail);
+
+        assertThat(mail.getMessage().getSubject()).isEqualTo("pre My subject");
+    }
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java]: Verify that a mail is sent

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+
+        redirect.init(mailetConfig);
+    }
+
+    @Test
+    public void initShouldNotThrowWhenEveryParameters() throws Exception {
+        FakeMailetConfig mailetConfig = FakeMailetConfig.builder()
+                .mailetName(MAILET_NAME)
+                .mailetContext(fakeMailContext)
+                .setProperty("static", "true")
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java]: Constants for property names PLZ

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                .name(MAILET_NAME)
+                .sender(MailAddressFixture.ANY_AT_JAMES)
+                .mimeMessage(mimeMessage)
+                .build();
+
+        redirect.service(mail);
+
+        SentMail newMail = fakeMailContext.getSentMails().get(0);
+        assertThat(newMail.getSubject()).contains("pre subj");
+    }
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java]: Check sent mails

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+        resend.init(mailetConfig);
+    }
+
+    @Test
+    public void initShouldNotThrowWhenEveryParameters() throws Exception {
+        FakeMailetConfig mailetConfig = FakeMailetConfig.builder()
+                .mailetName(MAILET_NAME)
+                .mailetContext(fakeMailContext)
+                .setProperty("debug", "true")
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java]: Constants

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+            newMail.getMessage().saveChanges();
+            newMail.removeAllAttributes();
+
+            if (senderDomainIsValid(newMail)) {
+                // Send it off...
+                mailet.getMailetContext().sendMail(newMail);
+            } else {
+                String logBuffer = mailet.getMailetName() + " mailet cannot forward " + originalMail.getName() + ". Invalid sender domain for " + newMail.getSender() + ". Consider using the Resend mailet " + "using a different sender.";
+                throw new MessagingException(logBuffer);
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: Too long !

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+            mailModifier.setRemoteAddr();
+            mailModifier.setRemoteHost();
+
+            if (isDebug) {
+                mailet.log("New mail - sender: " + newMail.getSender() + ", recipients: " + StringUtils.arrayToString(newMail.getRecipients().toArray()) + ", name: " + newMail.getName() + ", remoteHost: " + newMail.getRemoteHost() + ", remoteAddr: " + newMail.getRemoteAddr() + ", state: " + newMail.getState()
+                        + ", lastUpdated: " + newMail.getLastUpdated() + ", errorMessage: " + newMail.getErrorMessage());
+            }
+
+            // Create the message
+            boolean keepMessageId = false;
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: Re-affectation spotted.

In a keepMessageId() method : 

```
if (mailet.getInitParameters().getInLineType().equals(TypeCode.UNALTERED)) {
    return true;
}
return false;
```

In here you can keep the createAltered / createUnaltered stuff.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            mailModifier.setSubjectPrefix(originalMail);
+            mailModifier.setReplyTo(mailet.getReplyTo(originalMail), originalMail);
+            mailModifier.setReversePath(mailet.getReversePath(originalMail), originalMail);
+            mailModifier.setIsReply(mailet.getInitParameters().isReply(), originalMail);
+            mailModifier.setSender(mailet.getSender(originalMail), originalMail);
+            mailModifier.initializeDateIfNotPresent();
+            if (keepMessageId) {
+                mailModifier.setMessageId(originalMail);
+            }
+            newMail =  mailModifier.getMail();
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: Re-affectation

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+            newMail.getMessage().saveChanges();
+            newMail.removeAllAttributes();
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: Move newMail generation to a new method please

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+            if (senderDomainIsValid(newMail)) {
+                // Send it off...
+                mailet.getMailetContext().sendMail(newMail);
+            } else {
+                String logBuffer = mailet.getMailetName() + " mailet cannot forward " + originalMail.getName() + ". Invalid sender domain for " + newMail.getSender() + ". Consider using the Resend mailet " + "using a different sender.";
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: Get a ride of log buffer plz

**Comment by chibenwa** [comment]
Also resolve conficts, please

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java (hunk 11 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/redirect/ProcessRedirectNotify.java]: Not needed

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMailAttributeTest.java [hunk 14]
```diff
@@ -92,18 +97,4 @@ public void shouldOverwriteAttributeWhenAttributeAlreadyPresent() throws Messagi
 
         assertThat(mail.getAttribute("org.apache.james.junit1")).isEqualTo("bar");
     }
-
```
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/SetMailAttributeTest.java]: Unecessary blank line.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java (hunk 1 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: Same test with such configuration, and not so simple to achieve.
Testing sent mail should be an integration test.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java (hunk 2 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: For this message and the following:
i think the good way is to implement integration tests on those mailets.
This task should be done in another issue, as this one is already tl;dr

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java (hunk 3 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java]: same

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java (hunk 5 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java]: same

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java (hunk 6 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java]: same

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java (hunk 8 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java]: same

**Pull request closed without merging**